client/src/extension.ts

```diff
@@ -52,7 +52,14 @@ export function activate(context: ExtensionContext) {
         ],
         synchronize: {
             // Synchronize the setting section to the server.
-            configurationSection: ['python', 'pyright']
+            configurationSection: ['python', 'pyright'],
+
+            // Send file system change events within the workspace.
+            fileEvents: [
+                workspace.createFileSystemWatcher('**/pyrightconfig.json'),
+                workspace.createFileSystemWatcher('**/mspythonconfig.json'),
+                workspace.createFileSystemWatcher('**/*.{py,pyi,pyd}')
```

Comment:
I think these are the pre-defined ones being passed in the initialization unconditionally. Will this be overzealous? Or do we need to do this later once we know the search paths so we're able to exclude "bad" paths (other virtualenvs, for example) which won't be used?

I haven't tested this PR yet (I may do so tomorrow), but was wondering.

Reply:
Oh, this is in the client... I'll have to go read into what this does on this side of the equation.

Reply:
I assume the client installs a file watcher for the entire workspace (which makes sense, since it needs to update the file hierarchy in the editor). 

There's no way to omit "bad" paths in any of the three file watchers. You need to install the file watcher and ignore any events that you aren't interested in.

Reply:
I would think that it's an implementation detail of the server to be requesting file watches, not of the client. The above reads to me like the client forcing the watching of specific files onto the server (which may know better). A client which is not the pyright extension will not have this code.

It's more typical for the server to do this itself on handling of the `initialized` notification (which is when the client says "it's okay for you to send requests to me now", as it's illegal to make requests during the `initialize` handshake). See https://microsoft.github.io/language-server-protocol/specifications/specification-current/#initialized ("The server can use the initialized notification for example to dynamically register capabilities" -- file watching is a dynamically registered capability)

Reply:
Do you know of any code samples that show how to use dynamically registered capabilities — preferably for file watcher events? I couldn't find any. 

BTW, I looked in the debugger at the client capabilities sent from VS Code to the server, and they don't include the capability for file watcher dynamic registrations.

Reply:
Here is what the Go language server does: https://github.com/golang/tools/blob/8cb32c46614565a65efe53a937bd4144a4642627/internal/lsp/general.go#L143

Note that they only register a single watcher and give it a statically generated name; you'll want to create some unique identifier as you'll need more than one. The API is nearly identical to the client side version.

Reply:
> BTW, I looked in the debugger at the client capabilities sent from VS Code to the server, and they don't include the capability for file watcher dynamic registrations.

You mean that VS Code claims to not support file watching at all? That's strange... and goes against my experience so far. Maybe there's some option being missed as lacking file watch support is a biggie.